src/web100srv.c

```diff
@@ -1527,647 +1382,422 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
-                    "exceeded wait time",
-                    head_ptr->pid, head_ptr->pipe);
-        // Should send new 9977 'test aborted' signal to client.  Using
-        // this for now
-        // rac 3/26/10
-
-        log_println(6, "pid=%d, client='%s', stime=%ld, qtime=%ld now=%ld",
-                    head_ptr->pid, head_ptr->addr, head_ptr->stime,
-                    head_ptr->qtime, time(0));
-        log_println(6, "pipe-fd=%d, running=%d", head_ptr->pipe,
-                    head_ptr->running);
-        // boot the client
-        child_signal = SRV_QUEUE_SERVER_BUSY;
-        write(head_ptr->pipe, &child_signal, sizeof(int));
-        close(head_ptr->pipe);
-        tpid = head_ptr->pid;
-        child_sig(-1);
-        kill(tpid, SIGTERM);
-        child_sig(tpid);
-        continue;
-      }
-    }
+/**
+ * Enqueues a new child onto the list.
+ * @param head The head of the list
+ * @param new_child The child to enqueue
+ */
+void enqueue_child(ndtchild **head, ndtchild *new_child) {
+  ndtchild *current;
+  if (*head == NULL) {
+    // Enqueue onto the empty list
+    *head = new_child;
+  } else {
+    // Enqueue onto the end of a non-empty list
+    current = *head;
+    while (current->next != NULL) current = current->next;
+    current->next = new_child;
+  }
+}
 
-    if ((multiple == 1) && (mclients > waiting)) {
-      log_println(
-          5,
-          "Multi-client mode has uncaught terminated clients mclient=%d, "
-          "waiting=%d",
-          mclients, waiting);
-      while (mclients > waiting) {
-        child_sig(0);
-        mclients--;
-        log_println(5, "Removed terminated client from count mclients=%d",
-                    mclients);
-      }
-    }
+/**
+ * Attempts to enqueue a new ndtchild struct onto the queue of clients to be
+ * run. This will not enqueue the child unless the server has spare capacity.
+ * This function takes ownership of the memory pointed to by new_child, and
+ * either frees that memory or passes ownership to the linked list that starts
+ * at *head.
+ * @param new_child The new client
+ * @param head The head of the client queue
+ */
+void attempt_enqueue(ndtchild *new_child, ndtchild **head) {
+  ParentChildSignal child_signal;
+  ndtchild *current;
+  int client_count = 0;
+  // Count the number of connected clients
+  for (current = *head; current != NULL; current = current->next) {
+    client_count++;
+  }
+
+  if (!server_is_overloaded(client_count)) {
+    // The server is not overloaded, so queue up the new client.
+    enqueue_child(head, new_child);
+  } else {
+    // The server is overloaded. Reject the client and discard it.
+    log_println(0,
+                "Too many clients/mclients (%d) waiting to be served, "
+                "Please try again later.",
+                new_child->pid);
+    child_signal = SRV_QUEUE_SERVER_BUSY;
+    write(new_child->pipe, &child_signal, sizeof(ParentChildSignal));
```

Comment:
Can we have a wrapper function like:

`send_message_to_child(int child_pipe, ParentChildSignal message)`

It means that instead of this:

```
ParentChildSignal child_signal;
child_signal = SRV_QUEUE_SERVER_BUSY;
write(new_child->pipe, &child_signal, sizeof(ParentChildSignal));
```

We could just do this:

```
send_message_to_child(new_child->pipe, SRV_QUEUE_SERVER_BUSY);
```